server/src/achievement/achievement.service.ts

```diff
@@ -194,4 +198,153 @@ export class AchievementService {
       achievementType: ach.achievementType as AchievementTypeDto,
     };
   }
+
+  /** AchievementTracker functions */
+
+  /** Creates an achievement tracker */
+  async createAchievementTracker(user: User, achievementId: string) {
+    const existing = await this.prisma.achievementTracker.findFirst({
+      where: { userId: user.id, achievementId },
+    });
+
+    if (existing) {
+      return existing;
+    }
+
+    const progress = await this.prisma.achievementTracker.create({
+      data: {
+        userId: user.id,
+        progress: 0,
+        achievementId,
+      },
+    });
+
+    return progress;
+  }
+
+  async getAchievementTrackerByAchievementId(
+    user: User,
+    achievementId: string,
+  ) {
+    return await this.prisma.achievementTracker.findFirst({
+      where: { userId: user.id, achievementId },
+    });
+  }
+
+  async dtoForAchievementTracker(
+    tracker: AchievementTracker,
+  ): Promise<AchievementTrackerDto> {
+    const achievement = await this.getAchievementFromId(tracker.achievementId);
+    return {
+      userId: tracker.userId,
+      progress: tracker.progress,
+      achievementId: tracker.achievementId,
+      dateComplete: tracker.dateComplete?.toISOString(),
+    };
+  }
+
+  /** Emits & updates an achievement tracker */
+  async emitUpdateAchievementTracker(
+    tracker: AchievementTracker,
+    target?: User,
+  ) {
+    const dto = await this.dtoForAchievementTracker(tracker);
+
+    await this.clientService.sendProtected(
+      'updateAchievementTrackerData',
+      target?.id ?? tracker.userId,
+      dto,
+      {
+        id: dto.achievementId,
+        subject: 'AchievementTracker',
+        prismaStore: this.prisma.achievementTracker,
+      },
+    );
+  }
+
+  /** checks for all achievements associated with a user for a given completed challenge. */
+  async checkAchievementProgress(
+    user: User,
+    challengeId: string,
+    isJourney: boolean,
+  ) {
+    // find challenge corresponding to challengeId
+    const curChallenge = await this.prisma.challenge.findUniqueOrThrow({
+      where: { id: challengeId },
+    });
+
+    const ability = await this.abilityFactory.createForUser(user);
+
+    // find all achievements associated with the challenge that are accessible
+    // by user and have incomplete trackers; joins tracker to resulting query
+    const achs = await this.prisma.achievement.findMany({
+      where: {
+        OR: [
+          { linkedEventId: challengeId }, // achievements linked to the specific event of the challenge
+          { linkedEventId: null }, // achievements not linked to any specific event
+        ],
```

Comment:
I like where this is going, make sure to include the various enums like achievement type and locationtype = ANY